src/core/qgstiledownloadmanager.h

```diff
@@ -0,0 +1,225 @@
+{
+    Q_OBJECT
+  public:
+    QgsTileDownloadManagerReply( const QNetworkRequest &request ): mRequest( request ) {}
+    ~QgsTileDownloadManagerReply();
+
+    QByteArray data() const { return mData; }
+
+    QNetworkRequest request() const { return mRequest; }
+
+  public slots:
+    void requestFinished( QByteArray data );
+
+  signals:
+    void finished();
+
+  private:
+    QNetworkRequest mRequest;
+    QByteArray mData;
+};
+
+
+/// @cond PRIVATE
+
+/**
+ * \ingroup core
+ * Reply object that is used in the worker thread of the tile download manager.
+ * \note This class is not a part of public API
+ */
+class QgsTileDownloadManagerReplyWorkerObject : public QObject
+{
+    Q_OBJECT
+  public:
+    QgsTileDownloadManagerReplyWorkerObject( const QNetworkRequest &request ): mRequest( request ) {}
+
+  public slots:
+    void replyFinished();
+
+  signals:
+    void finished( QByteArray data );
+
+  private:
+    QNetworkRequest mRequest;
+};
+
+
+/**
+ * \ingroup core
+ * Worker object that is used in the worker thread of the tile download manager.
+ * \note This class is not a part of public API
+ */
+class QgsTileDownloadManagerWorker : public QObject
+{
+    Q_OBJECT
+
+  public:
+    QgsTileDownloadManagerWorker( QObject *parent = nullptr ): QObject( parent ) {}
+    //~QgsTileDownloadManagerWorker();
+
+  public slots:
+    void queueUpdated();
+
+  signals:
+    void requestFinished( QString url, QByteArray data );
+
+  private:
+};
+
+/// @endcond
+
+
+
+/**
+ * \ingroup core
+ *
+ * Tile download manager handles downloads of map tiles for the purpose of map rendering.
+ * The purpose of this class is to handle a couple of situations that may happen:
+ *
+ * - a map rendering job starts which requests tiles from a remote server, then in a short
+ *   while user zooms/pans map, which would normally mean that all pending requests get
+ *   aborted and then restarted soon afterwards. The download manager lets the requests
+ *   to finish to avoid excessive load on servers and needless aborts and repeated requests.
+ * - multiple map rendering jobs start at a similar time, requesting map the same map tiles.
+ *   Normally they could be requested multiple times from the server - the download manager
+ *   groups these requests and only does a single request at a time.
+ *
+ * At this point, it is not recommended to use this class for other scenarios than map
+ * rendering: using it elsewhere could slow down map rendering or have some unexpected
+ * negative effects.
+ *
+ * How do things work:
+ *
+ * - Upon a request, a QgsTileDownloadManagerReply object (based on QObject) is returned,
+ *   encapsulating the pending reply. Client can wait for its finished() signal - when
+ *   it gets emitted, the request has finished with success or failure. Client can delete
+ *   the reply object before the request is processed - download manager will finish its
+ *   download (and it will get cached for a future use).
```

Comment:
Question -- how would a renderer operating on the main thread (eg layout map exports) safely use this? From my interpretation this process implies use of an event loop to wait until the finished signal can be emitted, which isn't safe to do in this circumstance.

Reply:
I am not sure there is a safe way to have renderers run in the main thread and doing some async network requests at the same time. As you say, that implies having an event loop to wait for signals, and other GUI events could cause havoc.

The tile download manager is nearly a drop-in replacement for direct use of QNetworkRequest - the behavior is meant to be the same with the finished() signal, so there should be no change in this regard (not making things better nor worse).

I guess to make things safe one would need to use some other means of thread synchronization instead of signals and event loops, or some horrible hacks like adding an event filter on the embedded event loop, "stealing" the events we don't want to deal with, and then posting them back after the event loop has finished :-)

Reply:
Well, there's the QgsBlockingNetworkRequest class which was designed for use by providers to handle this situation gracefully (it uses no event loops on the main thread). 

This change doesn't make the situation any worse regarding this issue, but it also doesn't help and at some point in the future I think we're going to have to address this properly and ban all event loops/processEvent calls from the data providers. That's why I'd have loved to see this considered in the design on the tile cache...